src/s2/s2polygon.cc

```diff
@@ -1282,41 +1291,48 @@ unique_ptr<S2Polygon> S2Polygon::DestructiveApproxUnion(
 }
 
 unique_ptr<S2Polygon> S2Polygon::DestructiveUnion(
-    vector<std::unique_ptr<S2Polygon>> polygons,
+    vector<unique_ptr<S2Polygon>> polygons,
     const S2Builder::SnapFunction& snap_function) {
-  // Effectively create a priority queue of polygons in order of number of
-  // vertices.  Repeatedly union the two smallest polygons and add the result
-  // to the queue until we have a single polygon to return.
-  using QueueType = std::multimap<int, unique_ptr<S2Polygon>>;
-  QueueType queue;  // Map from # of vertices to polygon.
-  for (auto& polygon : polygons)
-    queue.insert(std::make_pair(polygon->num_vertices(), std::move(polygon)));
+  if (polygons.empty())
+    return make_unique<S2Polygon>();
+
+  // Create a priority queue of polygons in order of number of vertices.
+  // Repeatedly union the two smallest polygons and add the result to the
+  // queue until we have a single polygon to return.  We use the index
+  // into `polygons` as the second component of the pair instead of a
+  // pointer so the comparison is deterministic rather than depending
+  // on the polygon's address.  (Some clients depend on determinism.)
+  using Pair = pair<int, size_t>;  // (num vertices, `polygons` index)
+  using PriorityQueue = priority_queue<Pair, vector<Pair>, greater<Pair>>;
+  PriorityQueue queue;  // Least vertices is `top`.
+  for (size_t i = 0; i < polygons.size(); ++i) {
+    queue.emplace(polygons[i]->num_vertices(), i);
```

Comment:
Size of `polygons` doesn't change here, so you can use address instead of compute it in loop with index

Reply:
That makes the resulting polygon depend on the addresses of the input polygons, which some clients complained about.

Reply:
Hmm I probably should looked how union polygons work.
Ok

Reply:
Hmm I understand, above exist comment, I missed it

Reply:
But wait, addresses of polygon in vector is monotonically increasing.
So I think this comment was about elements itself (addresses of polygons, not their cell in vector)

Reply:
I see. Yeah, the comment was about `S2Polygon *`. You want to use a `unique_ptr<S2Polygon> *`?  Yeah, that's probably better since it can't be confused with num_vertices.

Reply:
We'll probably do this in a future release.